### ISSUE TITLE
Ability to run a Reproducible Build comparison Jenkins job to Verify two builds are identical

### ISSUE THREAD

**Issue by andrew-m-leonard** [adoptium/ci-jenkins-pipelines#301]
To aid verification of reproducible builds, it would be nice to have a Jenkins job that takes parameters like:
- JDK 1 download "URL"
- JDK 2 download "URL"
- Platform (eg.MacOS_aarch64)

The job then downloads both JDKs and performs a strict reproducible build comparison on them to indicate if they are identical.
"Platform" is required as reproducible build verification will vary based on platform (eg.Mac binaries need signing Signature removing before compare).

**Comment by zdtsw** [comment]
i do not quite follow this issue. 
a jenkins job to download same JDK versions but from different URL(?) then use them build same code respectively and compare the outcome?

**Comment by zdtsw** [comment]
in the case of URL1 and URL2, do you have one example? 
I can only think of temurin*-binaries git repo.  When the upstream code has no change so the nightly build could produce the same binary.

**Comment by andrew-m-leonard** [comment]
So reproducible builds will be "RELEASE=true" type builds, as that enables the --with-source-date=version and the reproducible flag.
So doing a release build twice for example, the URL1 could be to the binaries repo JDK.tar.gz and URL2 could be a re-build of the same release in a Jenkins job artifact.
Equally it could be for test/debug purpose so URL could be to where ever you place your JDK.tar.gz, eg.google docs

**Comment by smlambert** [comment]
So this job will take 2 parameters, URL1, URL2 both of which would be expected to point to a build that has been run with RELEASE=true (but this will be an expectation that the user knows this and/or reads the fine print under the parameter.

As an additional note, it is similar to how test jobs can take CUSTOMIZED_SDK_URL, this includes if you have build something locally on your laptop and then use UploadFile job to host it, to pass the URL1 or URL2 to this job.

**Comment by zdtsw** [comment]
Sum up after talk  with @andrew-m-leonard :
- release(reproducible) build is only applied on jdk19 (and will be back ported to jdk17 later), focus now is on jdk19(head) builds
- in reality we want to build twice on the same SHA from upstream(openjdk source) which generate two builds' artifacts correspondingly for downloadable as URL1 and URL2
- Preferably this should be a new dedicated jenkins job, which we can run manually when needed but mostly be triggered as part of the  weekly-openjdkX-pipeline in the future
- At the moment, we have `weekly-openjdk19-pipeline` triggers `openjdk19-pipeline` on hotspot and on temurin (at the same time) then each triggers downstream build jobs on different OS*platform. 
the artifacts from downstream build jobs are archived to themselves and populated to `openjdk19-pipeline` but not `weekly-openjdk19-pipeline` Due to the fact we only store the lastSuccessfulBuild artifact, we cannot get artifacts from multiple builds from jenkins ( build1 for URL1 is gone when build2 for URL2 is done).
-  thoughts for what to do next (break down steps):

1.   resolve: where and how to store these build artifacts from build1 and build2?
2.   be able to have different URLs so can do manual test(comparison)
3.   create jenkins job to automate manual test
4.   decide how we want to change for "build twice": in openjdk19-pipeline with new flag or in weekly-openjdk19-pipeline
5.   chain this job into weekly-openjdk19-pipeline

  

**Question** : do we need to store artifacts in the downstream build job ( e.g: https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk/job/jdk-linux-aarch64-hotspot/), if it (as successful build) eventually populates to its caller  openjdkX-pipeline? if we can free up disk space from these build jobs, we can instead archive all artifacts in weekly-openjdkX-pipeline(not only the lastSuccessfulBuild)

**Comment by zdtsw** [comment]
Ref: https://github.com/adoptium/temurin-build/issues/3015

**Comment by zdtsw** [comment]
test:

- positive(same build to test with itself)https://ci.adoptopenjdk.net/view/work-in-progress/job/wen_reproducible_comparision/8/
- negative(different build result one from x64 linux another from ppc64le) https://ci.adoptopenjdk.net/view/work-in-progress/job/wen_reproducible_comparision/9/console

**Comment by zdtsw** [comment]
to make it easier to track progress:
move this issue to only for Linux build
create top epic https://github.com/adoptium/ci-jenkins-pipelines/blob/master/tools/reproduce_comparison/Jenkinsfile#L51
and two new issue for MacOS and windows respectively

**Comment by zdtsw** [comment]
> Current the weekly run are copying all artifacts ( *msi, *.json, *sig, etc.). Might be worth to narrow down to required jdk jars? [#401 (comment)](https://github.com/adoptium/ci-jenkins-pipelines/pull/401#issuecomment-1238180809)

yes, i have a draft https://github.com/zdtsw/ci-jenkins-pipelines/commit/da40ea15634efffe426096baa6316d9e24af37a1 which i was trying to test it on weekly jdk19 and see if it gets all tar and zip files needed before making a PR to master.

**Comment by zdtsw** [comment]
created job : https://ci.adoptopenjdk.net/view/ReproducibleBuild/job/reproducible_comparision/
